### PR TITLE
Store and map projection progress

### PR DESCRIPTION
**Pull request by wwajerowicz**

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -30,6 +30,7 @@ val akkaHttpVersion                 = "10.1.7"
 val akkaPersistenceCassandraVersion = "0.92"
 val akkaPersistenceInMemVersion     = "2.5.1.1"
 val catsVersion                     = "1.6.0"
+val commonsVersion                  = "0.11.0"
```
Comment (umbreak) [inline build.sbt]: 0.11.1

File: modules/stream/src/main/scala/ch/epfl/bluebrain/nexus/sourcing/persistence/ProjectionProgress.scala [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+  /**
+    * The last processed [[Offset]].
+    */
+  def offset: Offset
+
+  /**
+    * Count of processed events.
+    */
+  def processedCount: Long
```
Comment (umbreak) [inline modules/stream/src/main/scala/ch/epfl/bluebrain/nexus/sourcing/persistence/ProjectionProgress.scala:17]: I guess we need also `discardedCount`, `evaluatedCount`

**Review by umbreak — CHANGES_REQUESTED** [review]

File: modules/stream/src/main/scala/ch/epfl/bluebrain/nexus/sourcing/StreamByTag.scala [hunk 3]
```diff
@@ -65,59 +69,95 @@ object StreamByTag {
...
-                  log.warning("Indexing event with id '{}' and value '{}' failed '{}'", id, event, "Mapping failed")
-                  None
-              }
-              .recoverWith(logError(id, event))
-            F.toIO(mapped).unsafeToFuture()
+          case (id, Some(event)) => mapEvent(id, event)
+          case (id, None)        => Future.successful((id, None, None))
+        }
+        .scan((initialProgress, None: Option[String], None: Option[Event], None: Option[MappedEvt])) {
+          (previous, next) =>
```
Comment (umbreak) [inline modules/stream/src/main/scala/ch/epfl/bluebrain/nexus/sourcing/StreamByTag.scala]: why not to use
```scala
case ((_, progress), next) =>
```

File: modules/stream/src/main/scala/ch/epfl/bluebrain/nexus/sourcing/StreamByTag.scala [hunk 4]
```diff
@@ -65,59 +69,95 @@ object StreamByTag {
...
+              case (envelope, eventOpt, mappedOpt) =>
+                (OffsetProgress(envelope.offset, progress.processedCount + 1, progress.discardedCount + 1),
+                 Some(envelope.persistenceId),
+                 eventOpt,
+                 mappedOpt)
+            }
+        }
+        .flatMapConcat {
+          case (progress, Some(id), eventOpt, mappedOpt) => Source.single((progress, id, eventOpt, mappedOpt))
+          case (_, None, _, _)                           => Source.empty
```
Comment (umbreak) [inline modules/stream/src/main/scala/ch/epfl/bluebrain/nexus/sourcing/StreamByTag.scala]: case _ =>

File: modules/stream/src/main/scala/ch/epfl/bluebrain/nexus/sourcing/persistence/IndexerConfig.scala [hunk 5]
```diff
@@ -92,6 +96,8 @@ object IndexerConfig {
       name: Option[String] = None,
       mapping: Option[Event => F[Option[MappedEvt]]] = None,
       index: Option[List[MappedEvt] => F[Unit]] = None,
+      mapInitialProgress: Option[ProjectionProgress => F[Unit]] = None,
```
Comment (umbreak) [inline modules/stream/src/main/scala/ch/epfl/bluebrain/nexus/sourcing/persistence/IndexerConfig.scala]: If in the end when building IndexerConfig you are defaulting to `_ => F.unit`, why not to do it here already:
```scala
mapInitialProgress: Option[ProjectionProgress => F[Unit]] = Some(_ => F.unit)
```

File: modules/stream/src/main/scala/ch/epfl/bluebrain/nexus/sourcing/persistence/IndexerConfig.scala [hunk 6]
```diff
@@ -92,6 +96,8 @@ object IndexerConfig {
...
+      mapProgress: Option[ProjectionProgress => F[Unit]] = None,
```
Comment (umbreak) [inline modules/stream/src/main/scala/ch/epfl/bluebrain/nexus/sourcing/persistence/IndexerConfig.scala]: same here

File: modules/stream/src/main/scala/ch/epfl/bluebrain/nexus/sourcing/persistence/ResumableProjection.scala [hunk 7]
```diff
@@ -25,43 +24,43 @@ trait ResumableProjection[F[_]] {
   def identifier: String
 
   /**
-    * @return the latest known offset; an inexistent offset is represented by [[akka.persistence.query.NoOffset]]
+    * @return the latest known [[ProjectionProgress]]; an inexistent offset is represented by [[ProjectionProgress]]
```
Comment (umbreak) [inline modules/stream/src/main/scala/ch/epfl/bluebrain/nexus/sourcing/persistence/ResumableProjection.scala]: an inexistent progress is represented by [[ProjectProgress.NoProgress]]

**Review by umbreak — CHANGES_REQUESTED** [review]

**Comment by umbreak** [comment]
Shouldn't an interface with the `StreamCoordinator` actor be implemented somewhere here, as:
```scala
// Calls the StreamCoordinator like: actor ? FetchLatestState
def fetchState[A]: F[Option[A]] = ???
// Calls the StreamCoordinator like: actor ? LatestState(Some(value))
def state[A](value: A): F[Unit] = ??
// Calls the StreamCoordinator like: actor ? Stop flatmap actor ? Done
def stop: F[Unit] = ??
```

File: modules/stream/src/main/scala/ch/epfl/bluebrain/nexus/sourcing/stream/StreamCoordinator.scala [hunk 8]
```diff
@@ -2,141 +2,197 @@ package ch.epfl.bluebrain.nexus.sourcing.stream
...
 
-    case FetchLatestState => sender() ! LatestState(state)
+  /**
+    * Fetches the latest state from the underlying actor [[StreamCoordinatorActor]] .
+    *
+    * @return latest state wrapped in [[F]]
+    */
+  def state(): F[Option[A]] = IO.fromFuture(IO(actor ? FetchLatestState)).to[F].map {
+    case st: LatestState[A] => st.state
   }
```
Comment (umbreak) [inline modules/stream/src/main/scala/ch/epfl/bluebrain/nexus/sourcing/stream/StreamCoordinator.scala:38]:
```scala
.recoverWith {
      case th: TimeoutException =>
        log.error(th, "Timed out while fetching state")
        F.raiseError(...)
      case NonFatal(th) =>
        log.error(th, s"Error while fetching state")
        F.raiseError(...)
    }
```
And maybe we need some `StreamCoordinator` errors

**Review by umbreak — CHANGES_REQUESTED** [review]

File: modules/stream/src/main/scala/ch/epfl/bluebrain/nexus/sourcing/stream/StreamCoordinator.scala [hunk 9]
```diff
@@ -2,141 +2,197 @@ package ch.epfl.bluebrain.nexus.sourcing.stream
...
-    // $COVERAGE-ON$
+  private implicit val askTimeout: Timeout = config.askTimeout
 
-    case FetchLatestState => sender() ! LatestState(state)
+  /**
+    * Fetches the latest state from the underlying actor [[StreamCoordinatorActor]] .
+    *
+    * @return latest state wrapped in [[F]]
+    */
+  def state(): F[Option[A]] = IO.fromFuture(IO(actor ? FetchLatestState)).to[F].map {
```
Comment (umbreak) [inline modules/stream/src/main/scala/ch/epfl/bluebrain/nexus/sourcing/stream/StreamCoordinator.scala]: `ClassTag` of A in order to pattern match on the value of A
```scala
def state()(implicit A: ClassTag[A]): F[Option[A]] = IO.fromFuture(IO(actor ? FetchLatestState)).to[F].map {
   case LatestState(A(state)) => state
}
```

**Review by umbreak — CHANGES_REQUESTED** [review]

File: modules/stream/src/main/scala/ch/epfl/bluebrain/nexus/sourcing/stream/StreamCoordinator.scala (hunk 8 above)
Comment (wwajerowicz) [reply modules/stream/src/main/scala/ch/epfl/bluebrain/nexus/sourcing/stream/StreamCoordinator.scala:38]: won't we handle these error at a higher level in KG and `InternalErrors`?

Comment (umbreak) [reply modules/stream/src/main/scala/ch/epfl/bluebrain/nexus/sourcing/stream/StreamCoordinator.scala:38]: In sourcing it is done this way. You can check in `AkkaAggregate` and `AkkaAggregateActor` in order to be consistent.

**Review by umbreak — CHANGES_REQUESTED** [review]

**Review by umbreak — APPROVED** [review]

**Pull request merged**